File: Squirrelssssasdq/mobile/nearby/android/src/main/java/app/squirrelsocial/nearby/NearbyProtocol.kt
package app.squirrelsocial.nearby

import java.util.UUID

/**
 * Wire constants shared with the backend (backend/nearby/protocol.py) and the iOS client.
 *
 * The advertisement carries ONLY [SERVICE_UUID]. The rotating anonymous id is read over GATT from
 * [BLE_ID_CHARACTERISTIC_UUID] as `[PAYLOAD_VERSION][16-byte id]` — iOS cannot put service or
 * manufacturer data in an advertisement, so both platforms use the same GATT read.
 */
object NearbyProtocol {
    val SERVICE_UUID: UUID = UUID.fromString("3c0b02ed-d244-4234-a411-8cdaf5812f97")
    val BLE_ID_CHARACTERISTIC_UUID: UUID = UUID.fromString("da2ec0af-a58d-4d28-aa65-ce93dfab8cbf")
    const val PAYLOAD_VERSION: Byte = 1
    const val BLE_ID_BYTES = 16

    /** Ids rotate on fixed wall-clock windows shared by every device. */
    const val ROTATION_MS = 15 * 60_000L
    const val UPLOAD_INTERVAL_MS = 60_000L
    const val MIN_SIGHTING_SPACING_MS = 5_000L
    const val MAX_SIGHTING_AGE_MS = 15 * 60_000L
    const val PEER_CACHE_MAX_MS = 5 * 60_000L
    const val MAX_UPLOAD_BATCH = 200

    fun encodePayload(bleIdHex: String): ByteArray {
        require(bleIdHex.length == BLE_ID_BYTES * 2) { "ble id must be $BLE_ID_BYTES bytes" }
        val out = ByteArray(1 + BLE_ID_BYTES)
        out[0] = PAYLOAD_VERSION
        for (i in 0 until BLE_ID_BYTES) {
            out[i + 1] = bleIdHex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
        return out
    }

    /** Returns the lowercase hex id, or null for anything that is not a valid Squirrel payload. */
    fun decodePayload(value: ByteArray?): String? {
        if (value == null || value.size != 1 + BLE_ID_BYTES || value[0] != PAYLOAD_VERSION) return null
        val sb = StringBuilder(BLE_ID_BYTES * 2)
        for (i in 1..BLE_ID_BYTES) sb.append(String.format("%02x", value[i].toInt() and 0xff))
        return sb.toString()
    }

    /** Start of the rotation window after [nowMs]; cached peer ids are invalid past this point. */
    fun nextRotationBoundary(nowMs: Long): Long = (nowMs / ROTATION_MS + 1) * ROTATION_MS
}

File: mobile/nearby/ios/SquirrelNearby/NearbyProtocol.swift
import CoreBluetooth
import Foundation

/// Wire constants shared with the backend (backend/nearby/protocol.py) and the Android client.
///
/// The advertisement carries ONLY `serviceUUID` (iOS cannot advertise service or manufacturer
/// data). The rotating anonymous id is read over GATT from `bleIdCharacteristicUUID` as
/// `[payloadVersion][16-byte id]`.
enum NearbyProtocol {
    static let serviceUUID = CBUUID(string: "3C0B02ED-D244-4234-A411-8CDAF5812F97")
    static let bleIdCharacteristicUUID = CBUUID(string: "DA2EC0AF-A58D-4D28-AA65-CE93DFAB8CBF")
    static let payloadVersion: UInt8 = 1
    static let bleIdBytes = 16

    /// Ids rotate on fixed wall-clock windows shared by every device.
    static let rotationSeconds: TimeInterval = 15 * 60
    static let uploadInterval: TimeInterval = 60
    static let minSightingSpacing: TimeInterval = 5
    static let maxSightingAge: TimeInterval = 15 * 60
    static let peerCacheMax: TimeInterval = 5 * 60
    static let maxUploadBatch = 200

    static func encodePayload(_ hex: String) -> Data? {
        guard hex.count == bleIdBytes * 2 else { return nil }
        var data = Data([payloadVersion])
        var index = hex.startIndex
        for _ in 0..<bleIdBytes {
            let next = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<next], radix: 16) else { return nil }
            data.append(byte)
            index = next
        }
        return data
    }

    /// Lowercase hex id, or nil for anything that is not a valid Squirrel payload.
    static func decodePayload(_ data: Data?) -> String? {
        guard let data, data.count == 1 + bleIdBytes, data.first == payloadVersion else { return nil }
        return data.dropFirst().map { String(format: "%02x", $0) }.joined()
    }

    /// Start of the rotation window after `date`; cached peer ids are invalid past this point.
    static func nextRotationBoundary(after date: Date) -> Date {
        let t = date.timeIntervalSince1970
        return Date(timeIntervalSince1970: (floor(t / rotationSeconds) + 1) * rotationSeconds)
    }
}
